Support Private Networks

guruchai opened the pull request.

Chaffelson commented on roles/infrastructure/defaults/main.yml:
```diff
@@ -28,6 +28,7 @@ infra__vpc_public_subnets_suffix:   "{{ common__vpc_public_subnets_suffix }}"
 
 # Infra
 infra__type:                        "{{ common__infra_type }}"
+infra__tunnel:                      "{{ common__tunnel }}"
```
What do we think about defaulting the tunnel to True? That seems the configuration more in line with our philosophy of 'working OOTB and as secure as possible' @wmudge

wmudge replied:
Only way we can allow this is if `public_endpoint_access` is available on GCP/Azure, right?

guruchai replied:
We could do this. In GCP/Azure, that option will be ignored. 
This will be the case for those providers even if you set it to true in definition file.

guruchai replied:
Wanted to make the switch once all the networking changes are in and we've tried it enough times.

Chaffelson replied:
Agreed. We can put the feature into this release, and make it the default once it has bedded in.

Chaffelson commented on roles/infrastructure/tasks/setup_aws_network.yml:
```diff
@@ -40,29 +40,27 @@
+    cidr: "{{ __aws_subnet_cidr_item }}"
     state: present
-    tags: "{{ infra__tags | combine(__aws_subnet_create_item.tags, recursive=True) }}"
+    tags: "{{ infra__tags | combine({ 'Name': '-'.join([infra__namespace, 'pub', __aws_subnet_index | string]) }, recursive=True) }}"
```
I would prefer if we didn't use a static 'pub' here and used a label variable like the others

guruchai replied:
let me fix this

Chaffelson commented on roles/infrastructure/tasks/setup_aws_network.yml:
```diff
@@ -82,10 +80,71 @@
+        vpc_id: "{{ infra__aws_vpc_id }}"
+        cidr: "{{ __aws_subnet_cidr_item }}"
+        state: present
+        tags: "{{ infra__tags | combine({ 'Name': '-'.join([infra__namespace, 'priv', __aws_subnet_index | string]) }, recursive=True) }}"
```
Same here with the label please

guruchai replied:
let me fix this

Chaffelson reviewed (changes requested):
Minor changes, will do more testing as well

guruchai commented:
@Dan, Addressed your comments

Chaffelson commented:
@guruchai can you please comment with the set of test cases for this? Which sets of variables should be set to what to test it out. 
I believe it's combinations of:
```
env:
  tunnel: yes
  public_endpoint_access: yes
```

L0: no, no
L1: yes, yes
L2: yes, no
?

wmudge commented on roles/infrastructure/defaults/main.yml:
```diff
@@ -91,7 +92,10 @@ infra__aws_vpc_az_count:            "{{ infra.aws.vpc.az_count | default(3) }}"
 infra__aws_igw_name:                "{{ infra.aws.vpc.internet_gateway.name | default([infra__namespace, infra__aws_igw_suffix] | join('-')) }}"
 infra__aws_igw_suffix:              "{{ infra.aws.vpc.internet_gateway.suffix | default(common__igw_suffix) }}"
 infra__aws_route_table_name:        "{{ infra.aws.vpc.labels.route_table | default([infra__namespace, infra__aws_route_table_suffix] | join('-')) }}"
```
For consistency, this variable should reflect that it is the _public_ route table, and a new variable added for a label, defaulting to "pub".

guruchai replied:
This is the default route table. We are leveraging it as public route table. Let me fix this.

wmudge commented on roles/infrastructure/defaults/main.yml:
```diff
@@ -91,7 +92,10 @@ infra__aws_vpc_az_count:            "{{ infra.aws.vpc.az_count | default(3) }}"
 infra__aws_igw_name:                "{{ infra.aws.vpc.internet_gateway.name | default([infra__namespace, infra__aws_igw_suffix] | join('-')) }}"
 infra__aws_igw_suffix:              "{{ infra.aws.vpc.internet_gateway.suffix | default(common__igw_suffix) }}"
 infra__aws_route_table_name:        "{{ infra.aws.vpc.labels.route_table | default([infra__namespace, infra__aws_route_table_suffix] | join('-')) }}"
+infra__aws_priv_route_table_name:   "{{ infra.aws.vpc.labels.route_table | default([infra__namespace, 'priv', infra__aws_route_table_suffix] | join('-')) }}"
```
Variable should be expanded to `private` from `priv`, as we are trying to be as clear as possible in our playbooks. (`ngw` is okay since the variable would get really long, but perhaps we should consider `gateway`?)

Again, for consistency, this needs its own variable, i.e. `infra.aws.vpc.labels.route_table_private`, and a label defaulting to "priv".

guruchai replied:
I am introducing `infra.aws.vpc.labels.public_route_table_suffix` and `infra.aws.vpc.labels.private_route_table_suffix` and getting rid off `infra__aws_route_table_suffix`.

wmudge commented on roles/infrastructure/defaults/main.yml:
```diff
@@ -91,7 +92,10 @@ infra__aws_vpc_az_count:            "{{ infra.aws.vpc.az_count | default(3) }}"
 infra__aws_route_table_name:        "{{ infra.aws.vpc.labels.route_table | default([infra__namespace, infra__aws_route_table_suffix] | join('-')) }}"
+infra__aws_priv_route_table_name:   "{{ infra.aws.vpc.labels.route_table | default([infra__namespace, 'priv', infra__aws_route_table_suffix] | join('-')) }}"
 infra__aws_route_table_suffix:      "{{ infra.aws.vpc.route_table.suffix | default(common__route_table_suffix) }}"
+infra__aws_ngw_name:                "{{ infra.aws.vpc.nat_gateway.name | default([infra__namespace, infra__aws_ngw_suffix] | join('-')) }}"
```
Hmm, perhaps expand to `infra__aws_nat_gateway` to align better with the configuration variable?

guruchai replied:
ngw is relatable to igw used elsewhere.

wmudge commented on roles/infrastructure/defaults/main.yml:
```diff
@@ -91,7 +92,10 @@ infra__aws_vpc_az_count:            "{{ infra.aws.vpc.az_count | default(3) }}"
+infra__aws_priv_route_table_name:   "{{ infra.aws.vpc.labels.route_table | default([infra__namespace, 'priv', infra__aws_route_table_suffix] | join('-')) }}"
 infra__aws_route_table_suffix:      "{{ infra.aws.vpc.route_table.suffix | default(common__route_table_suffix) }}"
+infra__aws_ngw_name:                "{{ infra.aws.vpc.nat_gateway.name | default([infra__namespace, infra__aws_ngw_suffix] | join('-')) }}"
+infra__aws_ngw_suffix:              "{{ infra.aws.vpc.nat_gateway.suffix | default(common__ngw_suffix) }}"
```
See above

wmudge commented on roles/infrastructure/tasks/setup_aws_network.yml:
```diff
@@ -40,29 +40,27 @@
+    infra__aws_public_subnet_ids: "{{ infra__aws_public_subnet_ids | default([]) | union([__aws_subnet_item.subnet.id | default('')]) }}"
   loop_control:
     loop_var: __aws_subnet_item
-    label: "{{ __aws_subnet_item.subnet.id }}"
```
Include a label -- produces better output within Ansible

guruchai replied:
I see. Thanks for tip!

wmudge commented on roles/infrastructure/tasks/setup_aws_network.yml:
```diff
@@ -82,10 +80,71 @@
+        az: "{{ __aws_az_info.availability_zones[__aws_subnet_index % infra__aws_vpc_az_count | int].zone_name }}"
+      loop_control:
+        loop_var: __aws_private_subnet_item
+        index_var: __aws_subnet_index
```
Include a loop label

wmudge commented on roles/infrastructure/tasks/teardown_aws_network.yml:
```diff
@@ -29,6 +29,36 @@
+            region: "{{ infra__region }}"
+            filters:
+              vpc-id: "{{ infra__aws_vpc_id }}"
+          register: __all_ngws
```
prefix with `aws`, i.e. `__aws_all_ngws`, for readability/debugging

wmudge commented on roles/platform/tasks/initialize_setup_aws.yml:
```diff
@@ -92,6 +92,12 @@
+- name: Set public subnets for public endpoint access
+  when: plat__public_endpoint_access
+  ansible.builtin.set_fact:
+    plat__aws_public_subnet_ids: "{{ infra__aws_public_subnet_ids }}"
```
Where is `infra__aws_public_subnet_ids` defined within the `plat` role?

wmudge replied:
Also, these are "internal" variables for the role, can you drop the `plat` prefix to show that they cannot be modified directly? (That's the intent, anyway, we'll see if we can uphold that goal!)

guruchai replied:
`infra__aws_public_subnet_ids` is a fact set in setup_aws_network.yml#L60. From what i read and verified, facts are shared across plays. 
I stick to the apparent pattern adopted by the code. Variables inside the play used across multiple files had the level prefixed to it like plat__region, plat__namespace, plat__aws_vpc_id, plat__aws_security_group_knox_id... Here the variable is further referenced in https://github.com/guruchai/cloudera.exe/blob/45721e98a5f8856cc0a7245f07e1fed840648a03/roles/platform/tasks/setup_aws_env.yml#L36. 

Let me know if `__aws_public_subnet_ids` is more conformant,  i can change is right away.

wmudge replied:
The pattern I'd like to follow is if the upstream "shared" variable is found, then it should be reassigned, like what you have here. If it is not found, because the upstream role has not been executed, e.g. the use of tags only executes the `platform` role, then this step should go out and discover the values.

Let's get this in and then post a PR to update with the discovery path.

guruchai commented:
@Chaffelson @wmudge Addressed your comments.

Changes:
1. `infra.aws.vpc.labels.route_table` is replaced with `infra.aws.vpc.labels.public_route_table` and `infra.aws.vpc.labels.private_route_table`. 
2. `globals.labels.route_table` is removed as it's specific to AWS. Instead I replaced `infra.aws.vpc.route_table.suffix` with `infra.aws.vpc.labels.public_route_table_suffix` and `infra.aws.vpc.labels.private_route_table_suffix` which default to 'public-rtb' and 'private-rtb' respectively.

Dan,
I added a few comments in the configuration doc. I have some example definition files for cldr-deploy which I will raise as PR once the existing PRs complete review. Let me know if you have different thoughts on examples.

wmudge commented on roles/runtime/defaults/main.yml:
```diff
@@ -25,6 +25,12 @@ run__namespace_cdp:                 "{{ common__namespace_cdp }}"
 run__vpc_name:                      "{{ common__vpc_name }}"
+run__vpc_public_subnets_suffix:     "{{ common__vpc_public_subnets_suffix }}"
+run__vpc_private_subnets_suffix:    "{{ common__vpc_private_subnets_suffix }}"
+run__vpc_public_subnets_name:       "{{ [run__namespace, run__vpc_public_subnets_suffix] | join('-') }}*"
```
Pattern should be a configuration variable, like `infra.vpc.public_subnets.name`, and then a default if that variable is not present in the inbound config.  Otherwise, variables like these two should go in `vars/main.yml`, as they are not "user serviceable parts."
